aspnet/mvc/controllers/routing.rst

```diff
@@ -1,8 +1,140 @@
+
+For most forms-over-data web applications that use standard controllers, conventional routes work nicely. Attribute routing is the recommended type of routing for use in APIs.
+
+Attribute routing
+-----------------
+
+[Route],[Http*]
+^^^^^^^^^^^^^^^
+
+Rather having to rely on naming and conventions to figre out where to route the request, attribute routing enables you to use an attribute to route requests directly to the desired destination. You may apply attributes to either controllers or actions. Attribute routing is more direct because routes must match what is defined in the attribute in order to be processed. When an action method is decorated with the ``[Route]`` attribute, it can no longer be accessed through the convention-based routes, as the attribute takes precedence. 
+
+.. code-block:: c#
+
+  [Route("category/{movie}.rss", Order = 0, Name = "MovieCategoryFeed")]
+
+A common scenario for using attribute routing in an API controller is having the same route behave differently depending upon which HTTP verb is used. For example, you might want a page that displays a movie by its id using the route ``/movies/1337``, but also on that page be able to purchase the movie, and send the results back to the server. To restrict the HTTP verb that is accepted, apply the ``[HttpGet]`` attribute to the ``GetMovie`` action:
+
+.. code-block:: c#
+
+  [HttpGet("{id}", Name = "GetMovie")]
+  public IActionResult GetMovie([FromRoute] int id)
+
+The order in which routes are processed is important, but you can control how it works. First, routing inspects the ``RouteOrder`` property of the ``[Route]`` attribute. 
+
+Routes on controller vs actions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Routes map URLs to actions in controllers. When at the action level, constraints may be applied to the actions to further restrict activity. 
+
+Inheritance
+^^^^^^^^^^^
+
+The ``IRouteConstraint`` interface exposes a ``Match`` method that you can use to create your own route constraints via inheritance, as the code below demonstrates:
+
+.. code-block:: c#
+
+  public class RouteConstraint : IRouteConstraint
+  {
+    public bool Match(HttpContext httpContext, 
+        IRouter route, string routeKey, 
+        IDictionary<string, object> values, 
+        RouteDirection routeDirection)
+        {
+          // code to build constraint
+        }
+  }
+
+Notice that the ``Match`` method contains parameters that you can use to inspect the route and HTTP request.
```

Comment:
need sample code. someplace in this doc you need to explain route data - would this be route data?